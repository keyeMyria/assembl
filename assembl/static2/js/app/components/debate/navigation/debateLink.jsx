// @flow
import React from 'react';
import { Link } from 'react-router';
import classNames from 'classnames';

import Timeline from './timeline';
import Slide from '../../common/transitions/slide';

type DebateLinkProps = {
  identifier: string,
  className: string,
  activeClassName: string,
  children: Array<*>,
  to: string,
  dataText: string,
  screenTooSmall: boolean
};

type DebateLinkState = {
  timelineActive: boolean
};

function DivContainer({ children, className }) {
  return <div className={className}>{children}</div>;
}

function SlideContainer({ children, className, active }) {
  return (
    <Slide in={active} className={className}>
      {children}
    </Slide>
  );
}

function TimelineContainer({ title, identifier, small, active, onMenuItemClick, onClose }) {
  const Container = small ? SlideContainer : DivContainer;
  return (
    <Container active={active} className={classNames('timeline-menu-container', { small: small })}>
      <section className={classNames('timeline-section', { vertical: small })} id="timeline">
        <div className="max-container">
          <Timeline title={title} vertical={small} identifier={identifier} onMenuItemClick={onMenuItemClick} onClose={onClose} />
        </div>
      </section>
    </Container>
  );
}

class DebateLink extends React.Component<*, DebateLinkProps, DebateLinkState> {
  state = {
    timelineActive: false
  };

  showTimeline = (event: SyntheticMouseEvent<>) => {
    this.setState({ timelineActive: true });
    if (this.props.screenTooSmall) event.preventDefault();
  };

  hideTimeline = () => {
    this.setState({ timelineActive: false });
  };

  render() {
    const { identifier, children, to, className, activeClassName, dataText, screenTooSmall } = this.props;
    const { timelineActive } = this.state;
    return (
      <div
        className={classNames('debate-link', { active: timelineActive })}
        onMouseOver={!screenTooSmall && this.showTimeline}
        onMouseLeave={!screenTooSmall && this.hideTimeline}
      >
        <Link
          to={to}
          className={classNames('debate-link-title', className, { selected: timelineActive })}
          activeClassName={activeClassName}
          data-text={dataText}
        >
          <span>{children}</span>
          {screenTooSmall && <span onClick={this.showTimeline} className="thumb-arrow assembl-icon assembl-icon-right-dir" />}
        </Link>
        <TimelineContainer
          title={children}
          identifier={identifier}
          small={screenTooSmall}
          active={timelineActive}
          onMenuItemClick={this.hideTimeline}
          onClose={this.hideTimeline}
        />
      </div>
    );
  }
}

export default DebateLink;