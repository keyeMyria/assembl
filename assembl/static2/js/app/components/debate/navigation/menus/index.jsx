// @flow
import React from 'react';
import TabMenuList from './tableMenuList';
import IndexMenuList from './indexMenuList';

export const MENU_KINDS = {
  index: 'index',
  table: 'table'
};

export type ItemNode = {
  id: string,
  title: string,
  img: {
    externalUrl: string
  },
  numContributors: number,
  numPosts: number,
  parentId: string
};

type MenuListProps = {
  kind: string
};

function MenuList(props: MenuListProps) {
  switch (props.kind) {
  case MENU_KINDS.index:
    return <IndexMenuList {...props} />;
  default:
    return <TabMenuList {...props} />;
  }
}

export { default as SurveyMenu } from './surveyMenu';
export { default as IdeasMenu } from './ideasMenu';
export { default as Menu } from './menu';
export default MenuList;