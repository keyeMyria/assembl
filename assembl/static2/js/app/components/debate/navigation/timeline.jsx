// @flow
import React from 'react';
import { connect } from 'react-redux';
import { I18n } from 'react-redux-i18n';
import classNames from 'classnames';

import { isCurrentPhase, getBarPercent, isStepCompleted, type Phase } from '../../../utils/timeline';
import TimelineSegment, { type DebateType } from './timelineSegment';
import TimelineMenu from './timelineMenu';
import Slide from '../../common/transitions/slide';

type TimelineProps = {
  title: string,
  debate: DebateType,
  identifier: string,
  vertical: boolean,
  locale: string,
  onMenuItemClick: Function,
  onClose: Function
};

type TimelineState = {
  menuItemOpen: boolean,
  selectedPhase: Phase | null
};

function TimeLineHeader({ title, currentPhase, onBack, locale }) {
  let phaseTitle = '';
  if (currentPhase) {
    currentPhase.title.entries.forEach((entry) => {
      if (locale === entry['@language']) {
        phaseTitle = entry.value;
      }
    });
  }
  return (
    <div className="timeline-header">
      <div className="menu-back-container">
        <div onClick={onBack} className="menu-back-btn">
          <span className="assembl-icon assembl-icon-left-small" />
          <span className="title">{I18n.t('debate.menuBack')}</span>
        </div>
      </div>

      <span className="title-container">
        <span>{title}</span>
        {phaseTitle && (
          <span className="phase-title">
            <span className="assembl-icon assembl-icon-right-dir" />
            <span>{phaseTitle}</span>
          </span>
        )}
      </span>
    </div>
  );
}

export class DumbTimeline extends React.Component<*, TimelineProps, TimelineState> {
  state = {
    menuItemOpen: false,
    selectedPhase: null
  };

  onShowMenu = (phase: Phase) => {
    this.setState({ selectedPhase: phase, menuItemOpen: true });
  };

  onHideMenu = () => {
    this.setState({ selectedPhase: null, menuItemOpen: false });
  };

  onBack = () => {
    const { onClose, vertical } = this.props;
    const { selectedPhase } = this.state;
    const hasCurrentPhase = !!selectedPhase;
    if (vertical) {
      if (hasCurrentPhase) {
        // first time we close the menu with her content
        // we need this for the slide transition ==> see onTimeLineMenuExited
        this.setState({ menuItemOpen: false });
      } else if (onClose) onClose();
    } else {
      this.setState({ selectedPhase: null, menuItemOpen: false }, () => {
        if (!hasCurrentPhase && onClose) onClose();
      });
    }
  };

  onTimeLineMenuExited = () => {
    // we remove the current phase
    this.setState({ selectedPhase: null });
  };

  render() {
    const { debateData } = this.props.debate;
    const { identifier, onMenuItemClick, vertical, title, locale } = this.props;
    const { selectedPhase, menuItemOpen } = this.state;
    return (
      <div className="timeline-panel">
        {vertical && <TimeLineHeader title={title} currentPhase={selectedPhase} locale={locale} onBack={this.onBack} />}
        <div className="timeline-body">
          <div className={classNames('timeline-container', { vertical: vertical })}>
            {debateData.timeline &&
              debateData.timeline.map((phase, index) => (
                <TimelineSegment
                  key={index}
                  index={index}
                  identifier={identifier}
                  phase={phase}
                  vertical={vertical}
                  isStepCompleted={isStepCompleted(phase)}
                  barPercent={getBarPercent(debateData.timeline[index])}
                  isCurrentPhase={isCurrentPhase(debateData.timeline[index])}
                  onMenuItemClick={onMenuItemClick}
                  onShowMenu={vertical && this.onShowMenu}
                  onHideMenu={vertical && this.onHideMenu}
                />
              ))}
          </div>
          {vertical && (
            <Slide in={menuItemOpen} className={'timeline-menu'} onExited={this.onTimeLineMenuExited}>
              {selectedPhase && <TimelineMenu vertical phase={selectedPhase} />}
            </Slide>
          )}
        </div>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  locale: state.i18n.locale,
  debate: state.debate
});

export default connect(mapStateToProps, null, null, { withRef: true })(DumbTimeline);