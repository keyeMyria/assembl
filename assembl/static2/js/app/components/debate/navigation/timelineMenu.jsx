// @flow
import React from 'react';
import { withApollo } from 'react-apollo';
import { Translate, Localize } from 'react-redux-i18n';
import { connect } from 'react-redux';

import Menu from './menus/menu';
import { getPhaseStatus, type Phase } from '../../../utils/timeline';
import { PHASE_STATUS } from '../../../constants';
import { MENU_KINDS } from './menus';

type TimelineMenuProps = {
  phase: Phase,
  locale: string,
  vertical: boolean,
  onMenuItemClick: Function
};

export class DumbTimelineMenu extends React.Component<*, TimelineMenuProps, void> {
  componentWillMount() {
    const { phase: { title, start, end }, locale } = this.props;
    this.phaseStatus = getPhaseStatus(start, end);
    let phaseName = '';
    title.entries.forEach((entry) => {
      if (locale === entry['@language']) {
        phaseName = entry.value.toLowerCase();
      }
    });
    this.phaseName = phaseName;
  }

  phaseStatus = null;

  phaseName = null;

  renderNotStarted = (className?: string) => {
    const { phase: { start } } = this.props;
    return (
      <div className={className}>
        <Translate value="debate.notStarted" phaseName={this.phaseName} />
        <Localize value={start} dateFormat="date.format" />
      </div>
    );
  };

  render() {
    const { phase: { identifier }, onMenuItemClick, vertical } = this.props;
    const isNotStarted = this.phaseStatus === PHASE_STATUS.notStarted;
    return (
      <div className="menu-container">
        {isNotStarted ? (
          this.renderNotStarted('not-started')
        ) : (
          <Menu identifier={identifier} onMenuItemClick={onMenuItemClick} kind={vertical ? MENU_KINDS.index : MENU_KINDS.table} />
        )}
      </div>
    );
  }
}

const mapStateToProps = state => ({
  locale: state.i18n.locale
});

// $FlowFixMe
export default connect(mapStateToProps)(withApollo(DumbTimelineMenu));