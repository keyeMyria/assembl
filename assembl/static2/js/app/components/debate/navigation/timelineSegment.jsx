// @flow
import React from 'react';
import { withApollo, type ApolloClient } from 'react-apollo';
import { browserHistory } from 'react-router';
import { Translate, Localize, I18n } from 'react-redux-i18n';
import { connect } from 'react-redux';
import classNames from 'classnames';

import { prefetchMenuQuery } from './menus/menu';
import { getPhaseStatus, isSeveralIdentifiers, type Timeline, type Phase } from '../../../utils/timeline';
import { displayModal } from '../../../utils/utilityManager';
import { get } from '../../../utils/routeMap';
import { PHASE_STATUS } from '../../../constants';
import TimelineMenu from './timelineMenu';

export type DebateType = {
  debateData: {
    timeline: Timeline,
    slug: string
  }
};

type TimelineSegmentProps = {
  index: number,
  client: ApolloClient,
  phase: Phase,
  debate: DebateType,
  barPercent: number,
  isCurrentPhase: boolean,
  isStepCompleted: boolean,
  locale: string,
  vertical: boolean,
  onMenuItemClick: Function,
  onShowMenu: Function,
  onHideMenu: Function
};

type TimelineSegmentState = {
  active: boolean
};

export class DumbTimelineSegment extends React.Component<*, TimelineSegmentProps, TimelineSegmentState> {
  state = {
    active: false
  };

  componentWillMount() {
    const { phase: { identifier, title, start, end }, locale, client } = this.props;
    this.phaseStatus = getPhaseStatus(start, end);
    let phaseName = '';
    let phaseTitle = '';
    title.entries.forEach((entry) => {
      if (locale === entry['@language']) {
        phaseTitle = entry.value;
        phaseName = entry.value.toLowerCase();
      }
    });
    this.phaseName = phaseName;
    this.phaseTitle = phaseTitle;
    prefetchMenuQuery(client, {
      lang: locale,
      identifier: identifier
    });
  }

  componentWillReceiveProps() {
    if (this.state.active) {
      this.setState({ active: false });
    }
  }

  phaseStatus = null;

  phaseName = null;

  phaseTitle = null;

  showMenu = () => {
    this.setState({ active: true }, () => {
      const { onShowMenu, phase } = this.props;
      if (onShowMenu) onShowMenu(phase);
    });
  };

  hideMenu = () => {
    this.setState({ active: false }, () => {
      const { onHideMenu } = this.props;
      if (onHideMenu) onHideMenu();
    });
  };

  renderNotStarted = (className?: string) => {
    const { phase: { start } } = this.props;
    return (
      <div className={className}>
        <Translate value="debate.notStarted" phaseName={this.phaseName} />
        <Localize value={start} dateFormat="date.format" />
      </div>
    );
  };

  displayPhase = () => {
    const { phase } = this.props;
    const { debateData } = this.props.debate;
    const isRedirectionToV1 = phase.interface_v1;
    const slug = { slug: debateData.slug };
    const params = { slug: debateData.slug, phase: phase.identifier };
    const isSeveralPhases = isSeveralIdentifiers(debateData.timeline);
    if (isSeveralPhases) {
      if (this.phaseStatus === PHASE_STATUS.notStarted) {
        const body = this.renderNotStarted();
        displayModal(null, body, true, null, null, true);
      }
      if (this.phaseStatus === PHASE_STATUS.inProgress || this.phaseStatus === PHASE_STATUS.completed) {
        if (!isRedirectionToV1) {
          browserHistory.push(get('debate', params));
          this.hideMenu();
        } else {
          const body = <Translate value="redirectToV1" phaseName={this.phaseName} />;
          const button = { link: get('oldDebate', slug), label: I18n.t('home.accessButton'), internalLink: false };
          displayModal(null, body, true, null, button, true);
          setTimeout(() => {
            window.location = get('oldDebate', slug);
          }, 6000);
        }
      }
    } else if (!isRedirectionToV1) {
      browserHistory.push(get('debate', params));
      this.hideMenu();
    } else {
      window.location = get('oldDebate', slug);
    }
  };

  render() {
    const { index, phase, onMenuItemClick, barPercent, isCurrentPhase, isStepCompleted, vertical } = this.props;
    const { active } = this.state;
    const timelineClass = classNames('timeline-title', {
      'txt-active-bold': vertical && isCurrentPhase,
      'txt-active-light': !vertical || (vertical && isStepCompleted),
      'txt-not-active': vertical && !isCurrentPhase && !isStepCompleted
    });
    const complitedOrCurrent = isStepCompleted || isCurrentPhase;
    const dimensionName = vertical ? 'height' : 'width';
    return (
      <div
        className={classNames('minimized-timeline', {
          active: !vertical && active
        })}
        onMouseOver={!vertical && this.showMenu}
        onMouseLeave={!vertical && this.hideMenu}
      >
        <div onClick={vertical ? this.showMenu : this.displayPhase} className={timelineClass}>
          <div className="timeline-link">{this.phaseTitle}</div>
          {vertical && (
            <div className="timeline-link-description">
              {isStepCompleted && I18n.t('debate.phaseCompleted')}
              {isCurrentPhase && I18n.t('debate.phaseStarted')}
              {!complitedOrCurrent && I18n.t('debate.phaseUpcoming')}
            </div>
          )}
        </div>
        <div className="timeline-graph">
          {vertical && (
            <div
              className={classNames('timeline-number', {
                active: complitedOrCurrent,
                'not-active': !complitedOrCurrent,
                'timeline-text': !isStepCompleted
              })}
            >
              {isStepCompleted ? (
                <span className="assembl-icon-checked white" />
              ) : (
                <span>
                  {isCurrentPhase && <span className="thumb-arrow assembl-icon assembl-icon-right-dir" />}
                  {index + 1}
                </span>
              )}
            </div>
          )}
          <div className="timeline-bars">
            {barPercent > 0 && (
              <div
                className="timeline-bar-filler"
                style={barPercent < 20 ? { [dimensionName]: '20%' } : { [dimensionName]: `${barPercent}%` }}
              >
                &nbsp;
              </div>
            )}
            <div className="timeline-bar-background">&nbsp;</div>
          </div>
        </div>
        {!vertical &&
          active && (
            <div>
              <span className="timeline-arrow" />
              <TimelineMenu phase={phase} onMenuItemClick={onMenuItemClick} />
            </div>
          )}
      </div>
    );
  }
}

const mapStateToProps = state => ({
  locale: state.i18n.locale,
  debate: state.debate
});

// $FlowFixMe
export default connect(mapStateToProps)(withApollo(DumbTimelineSegment));