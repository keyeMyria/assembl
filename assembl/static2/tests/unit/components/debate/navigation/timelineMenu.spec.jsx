import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';

import { DumbTimelineMenu } from '../../../../../js/app/components/debate/navigation/timelineMenu';

describe('TimelineMenu component', () => {
  it('should match the TimelineMenu', () => {
    const phase = {
      title: {
        entries: [{ en: 'Foo' }]
      },
      start: 'date1',
      end: 'date2',
      identifier: 'foo'
    };
    const props = {
      phase: phase,
      locale: 'en'
    };
    const renderer = new ShallowRenderer();
    renderer.render(<DumbTimelineMenu {...props} />);
    const rendered = renderer.getRenderOutput();
    expect(rendered).toMatchSnapshot();
  });
  it('should match the vetical TimelineMenu', () => {
    const phase = {
      title: {
        entries: [{ en: 'Foo' }]
      },
      start: 'date1',
      end: 'date2',
      identifier: 'foo'
    };
    const props = {
      phase: phase,
      vertical: true,
      locale: 'en'
    };
    const renderer = new ShallowRenderer();
    renderer.render(<DumbTimelineMenu {...props} />);
    const rendered = renderer.getRenderOutput();
    expect(rendered).toMatchSnapshot();
  });
});