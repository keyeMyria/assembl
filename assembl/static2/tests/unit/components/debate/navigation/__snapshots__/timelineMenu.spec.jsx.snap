// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimelineMenu component should match the TimelineMenu 1`] = `
<div
  className="menu-container"
>
  <Menu
    identifier="foo"
    kind="table"
    onMenuItemClick={undefined}
  />
</div>
`;

exports[`TimelineMenu component should match the vetical TimelineMenu 1`] = `
<div
  className="menu-container"
>
  <Menu
    identifier="foo"
    kind="index"
    onMenuItemClick={undefined}
  />
</div>
`;
